File: packages/companion/swift/Sources/BridgethingCompanion/VoiceDispatcher.swift
import BridgethingGateway
import BridgethingSchema
import Foundation

public protocol VoiceCapturing: Sendable {
    func start(gateway: BridgethingGateway) async
    func stop() async
    func setCatalogResolver(_ provider: @escaping @Sendable () async -> (any VoiceCatalogResolving)?) async
}

@available(macOS 26.0, iOS 26.0, *)
protocol NluSpeechRecognizing: Sendable {
    func prepare() async throws
    func transcribe(fileAt url: URL) async throws -> NluSpeechRecognizer.Transcription
}

@available(macOS 26.0, iOS 26.0, *)
extension NluSpeechRecognizer: NluSpeechRecognizing {}

protocol VoicePacketDecoding: Sendable {
    func decode(_ packets: [Data], format: VoiceFormat) throws -> Data
}

struct SystemVoicePacketDecoder: VoicePacketDecoding {
    func decode(_ packets: [Data], format: VoiceFormat) throws -> Data {
        try VoicePacketDecoder(format: format).decode(packets)
    }
}

@available(macOS 26.0, iOS 26.0, *)
public actor VoiceDispatcher: VoiceCapturing {
    private struct Capture {
        let format: VoiceFormat
        var packets: [UInt32: Data] = [:]
    }

    private struct Turn: Sendable {
        let streamId: UUID
        let format: VoiceFormat
        let packets: [Data]
    }

    private let recognizer: any NluSpeechRecognizing
    private let decoder: any VoicePacketDecoding
    private let controller: VoiceController
    private var resolverProvider: (@Sendable () async -> (any VoiceCatalogResolving)?)?
    private var captures: [UUID: Capture] = [:]
    private var tasks: [Task<Void, Never>] = []
    private var prewarmRequested = false

    public init(recognizer: NluSpeechRecognizer, controller: VoiceController) {
        self.init(recognizer: recognizer, decoder: SystemVoicePacketDecoder(), controller: controller)
    }

    init(
        recognizer: any NluSpeechRecognizing,
        decoder: any VoicePacketDecoding,
        controller: VoiceController
    ) {
        self.recognizer = recognizer
        self.decoder = decoder
        self.controller = controller
    }

    public func setCatalogResolver(_ provider: @escaping @Sendable () async -> (any VoiceCatalogResolving)?) {
        resolverProvider = provider
    }

    public func start(gateway: BridgethingGateway) async {
        tasks.append(Task { [recognizer] in
            do {
                try await recognizer.prepare()
            } catch {
                print("voice: speech recognizer unavailable: \(error)")
            }
        })
        let events = gateway.events
        tasks.append(Task { [weak self] in
            await withDiscardingTaskGroup { group in
                for await event in events {
                    guard case let .message(deviceId, message) = event,
                          case let .voice(surface) = message.data
                    else { continue }
                    guard let self else { return }
                    guard let turn = await self.accept(surface) else { continue }
                    group.addTask { await self.dispatch(turn, deviceId: deviceId, gateway: gateway) }
                }
            }
        })
    }

    public func stop() async {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
        captures.removeAll()
        prewarmRequested = false
    }

    private func accept(_ msg: BridgeToGatewayVoiceMsg) -> Turn? {
        switch msg {
        case let .streamOpen(open):
            self.open(open)
        case let .frame(frame):
            append(frame)
        case let .streamClose(close):
            return closeCapture(close)
        default:
            break
        }
        return nil
    }

    private func open(_ msg: VoiceStreamOpen) {
        captures[msg.streamId] = Capture(format: msg.format)
        guard !prewarmRequested else { return }
        prewarmRequested = true
        tasks.append(Task { [controller] in await controller.prewarm() })
    }

    private func append(_ msg: VoiceFrame) {
        captures[msg.streamId]?.packets[msg.seq] = msg.packet
    }

    private func closeCapture(_ msg: VoiceStreamClose) -> Turn? {
        guard let capture = captures.removeValue(forKey: msg.streamId) else { return nil }
        guard msg.reason == .endOfSpeech else { return nil }
        return Turn(
            streamId: msg.streamId,
            format: capture.format,
            packets: capture.packets.sorted { $0.key < $1.key }.map(\.value)
        )
    }

    private nonisolated func dispatch(_ turn: Turn, deviceId: String, gateway: BridgethingGateway) async {
        do {
            try await resolveAndDispatch(transcript: await transcribe(turn), deviceId: deviceId, gateway: gateway)
        } catch {
            print("voice: dispatching turn \(turn.streamId) failed: \(error)")
        }
    }

    private nonisolated func transcribe(_ turn: Turn) async -> String {
        guard !turn.packets.isEmpty else { return "" }
        do {
            let pcm = try decoder.decode(turn.packets, format: turn.format)
            guard !pcm.isEmpty else { return "" }
            let url = try Self.writeWav(pcm: pcm, format: turn.format)
            defer { try? FileManager.default.removeItem(at: url) }
            return try await recognizer.transcribe(fileAt: url).text
        } catch {
            print("voice: capture \(turn.streamId) failed: \(error)")
            return ""
        }
    }

    nonisolated func resolveAndDispatch(transcript: String, deviceId: String, gateway: BridgethingGateway) async throws {
        let resolution: VoiceController.Resolution
        do {
            resolution = try await controller.resolve(transcript: transcript)
        } catch {
            print("voice: nlu failed: \(error)")
            resolution = VoiceController.Resolution(
                resolved: NluPrediction(intent: NluIntentCatalog.noIntent, transcript: transcript).toWire(),
                stage: .noModel
            )
        }
        var prediction = NluPrediction.fromWire(resolution.resolved)

        if let resolver = await currentResolver() {
            do {
                prediction = try await resolver.decorate(prediction)
            } catch {
                print("voice: catalog resolution failed, dispatching without a uri: \(error)")
            }
        }

        try await gateway.device(deviceId).voice.dispatch(
            VoiceDispatch(resolved: prediction.toWire(), stage: resolution.stage.wire)
        )
    }

    private func currentResolver() async -> (any VoiceCatalogResolving)? {
        guard let resolverProvider else { return nil }
        return await resolverProvider()
    }

    // MARK: - wav

    static let wavBitsPerSample: UInt16 = 16

    static func writeWav(pcm: Data, format: VoiceFormat) throws -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("bridgething-voice-\(UUID().uuidString).wav")
        var out = Data()
        let channels = UInt16(format.channels)
        let bits = Self.wavBitsPerSample
        let rate = UInt32(format.sampleRateHz)
        let blockAlign = channels * bits / 8
        let byteRate = rate * UInt32(blockAlign)

        func append32(_ v: UInt32) { withUnsafeBytes(of: v.littleEndian) { out.append(contentsOf: $0) } }
        func append16(_ v: UInt16) { withUnsafeBytes(of: v.littleEndian) { out.append(contentsOf: $0) } }

        out.append(contentsOf: Array("RIFF".utf8))
        append32(UInt32(36 + pcm.count))
        out.append(contentsOf: Array("WAVEfmt ".utf8))
        append32(16)
        append16(1)
        append16(channels)
        append32(rate)
        append32(byteRate)
        append16(blockAlign)
        append16(bits)
        out.append(contentsOf: Array("data".utf8))
        append32(UInt32(pcm.count))
        out.append(pcm)

        try out.write(to: url)
        return url
    }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/VoiceDispatcher.kt
package com.bridgething.companion

import android.util.Log
import com.bridgething.gateway.BridgethingGateway
import com.bridgething.gateway.GatewayEvent
import com.bridgething.gateway.device
import com.bridgething.schema.BridgeToGatewayMsgData
import com.bridgething.schema.BridgeToGatewayVoiceMsg
import com.bridgething.schema.NluStage
import com.bridgething.schema.VoiceCloseReason
import com.bridgething.schema.VoiceDispatch
import com.bridgething.schema.VoiceFormat
import com.bridgething.schema.VoiceFrame
import com.bridgething.schema.VoiceStreamClose
import com.bridgething.schema.VoiceStreamOpen
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

public class VoiceDispatcher(
    private val recognizer: NluSpeechRecognizing,
    private val decoder: VoicePacketDecoding,
    private val controller: VoiceController,
) {
    private class Capture(val format: VoiceFormat) {
        val packets: ConcurrentHashMap<UInt, ByteArray> = ConcurrentHashMap()
    }

    private class Turn(val streamId: UUID, val format: VoiceFormat, val packets: List<ByteArray>)

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val mutex = Mutex()
    private val jobs = mutableListOf<Job>()
    private val captures = ConcurrentHashMap<UUID, Capture>()

    @Volatile
    private var resolverProvider: (suspend () -> VoiceCatalogResolving?)? = null

    @Volatile
    private var prewarmed = false

    public fun setCatalogResolver(provider: suspend () -> VoiceCatalogResolving?) {
        resolverProvider = provider
    }

    public suspend fun start(gateway: BridgethingGateway) {
        mutex.withLock {
            stopJobs()
            jobs.add(
                scope.launch {
                    runCatching { recognizer.prepare() }
                        .onFailure { Log.w(TAG, "speech recognizer unavailable: ${it.message ?: it}") }
                },
            )
            jobs.add(scope.launch { runCaptures(gateway) })
        }
    }

    public suspend fun stop() {
        mutex.withLock {
            stopJobs()
            captures.clear()
            prewarmed = false
        }
    }

    private fun stopJobs() {
        for (job in jobs) job.cancel()
        jobs.clear()
    }

    private suspend fun runCaptures(gateway: BridgethingGateway): Unit = coroutineScope {
        gateway.events.collect { event ->
            if (event !is GatewayEvent.Message) return@collect
            val surface = event.message.data as? BridgeToGatewayMsgData.Voice ?: return@collect
            when (val msg = surface.data) {
                is BridgeToGatewayVoiceMsg.StreamOpen -> open(msg.data, this)
                is BridgeToGatewayVoiceMsg.Frame -> append(msg.data)
                is BridgeToGatewayVoiceMsg.StreamClose -> {
                    val turn = closeCapture(msg.data)
                    if (turn != null) launch { dispatch(turn, event.deviceId, gateway) }
                }
                else -> Unit
            }
        }
    }

    private fun open(msg: VoiceStreamOpen, scope: CoroutineScope) {
        captures[msg.streamId] = Capture(msg.format)
        if (prewarmed) return
        prewarmed = true
        scope.launch {
            runCatching { controller.prewarm() }
                .onFailure { Log.w(TAG, "prewarm failed: ${it.message ?: it}") }
        }
    }

    private fun append(msg: VoiceFrame) {
        captures[msg.streamId]?.packets?.put(msg.seq, msg.packet)
    }

    private fun closeCapture(msg: VoiceStreamClose): Turn? {
        val capture = captures.remove(msg.streamId) ?: return null
        if (msg.reason != VoiceCloseReason.EndOfSpeech) return null
        return Turn(
            streamId = msg.streamId,
            format = capture.format,
            packets = capture.packets.entries.sortedBy { it.key }.map { it.value },
        )
    }

    private suspend fun dispatch(turn: Turn, deviceId: String, gateway: BridgethingGateway) {
        resolveAndDispatch(transcribe(turn), deviceId, gateway)
    }

    private suspend fun transcribe(turn: Turn): String {
        if (turn.packets.isEmpty()) return ""
        return try {
            val samples = decoder.decode(turn.packets, turn.format)
            if (samples.isEmpty()) "" else recognizer.transcribe(samples, turn.format.sampleRateHz.toInt())
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "capture ${turn.streamId} failed: ${e.message ?: e}")
            ""
        }
    }

    private suspend fun resolveAndDispatch(transcript: String, deviceId: String, gateway: BridgethingGateway) {
        val resolution = try {
            controller.resolve(transcript)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "nlu failed: ${e.message ?: e}")
            VoiceController.Resolution(
                NluPrediction(intent = NluIntentCatalog.NO_INTENT, transcript = transcript).toWire(),
                NluStage.NoModel,
            )
        }

        var prediction = NluPrediction.fromWire(resolution.resolved)
        val resolver = resolverProvider?.invoke()
        if (resolver != null) {
            prediction = try {
                resolver.decorate(prediction.copy(slots = prediction.slots.copy()))
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                Log.w(TAG, "catalog resolution failed, dispatching without a uri: ${e.message ?: e}")
                prediction
            }
        }

        gateway.device(deviceId).voice.dispatch(
            VoiceDispatch(resolved = prediction.toWire(), stage = resolution.stage),
        )
    }

    private companion object {
        const val TAG = "bridgething.voice"
    }
}
